File: October/13/buysellstocks.py
# LeetCode 121: Best Time to Buy and Sell Stock
# Explanation:
# 1. Track the minimum price seen so far.
# 2. At each price, compute potential profit = current_price - min_price.
# 3. Keep the maximum profit encountered.
# Time Complexity: O(n)
# Space Complexity: O(1)

class Solution:
    def maxProfit(self, prices):
        min_price = float('inf')
        max_profit = 0
        for p in prices:
            min_price = min(min_price, p)
            max_profit = max(max_profit, p - min_price)
        return max_profit

File: October/13/buysellstocks.js
// LeetCode 121: Best Time to Buy and Sell Stock
// Explanation:
// 1. Track minimum price seen so far.
// 2. Update max profit as we go.
// Time Complexity: O(n)
// Space Complexity: O(1)

var maxProfit = function(prices) {
    let minPrice = Infinity;
    let maxProfit = 0;
    for (let price of prices) {
        minPrice = Math.min(minPrice, price);
        maxProfit = Math.max(maxProfit, price - minPrice);
    }
    return maxProfit;
};

console.log(maxProfit([7,1,5,3,6,4])); // Output: 5
console.log(maxProfit([7,6,4,3,1]));   // Output: 0
